### PR TITLE
Dungeon: Improve dungeon speed with flash

### LINKED ISSUE #124: Improve dungeon clearance algorithm
### Is your feature request related to a problem? Please describe.
Currently, after we unlock the preview tiles, the dungeon pathing is going through all the tiles even tho we know they are not the boss.

### Describe the solution you'd like
Initial steps:
image

Dungeon End:
image

I think this would speed up the dungeon completeness by a lot, specially on later areas with a lot of HP and a lot of possible fights

### PR DESCRIPTION
**Pull request by Daniferrito**
A image is worth a thousand words, and a video is worth a thousand images, so... one million words?

Dungeon clearing gif
(Any small pause in the exploration is a fight)

Fixes #124 as per the last sugestion:

* Visit the non-enemy tile that will reveal the most amount of tiles (min 1)
* Visit any enemy tile that will reveal the most amount of tiles (min 1)

Still have to polish how the switches interact with the exploration. Probably we will need a third switch to toggle whether to kill all non-bosses (or reword the switches?).

It does nothing if flash is not unlocked for the dungeon.

### CODE REVIEW

**Comment by lmikulla** [comment]
When you have flash for dungeons you can move into all empty or treasure tiles (without opening chests) for free for extra exploration, avoiding unnecessary fights. With the current ticks speeds you can explore many more tiles that way (especially in large dungeons). 

Its worth updating the logic to go into ALL reachable non-combat tiles immediately, before picking the potentially best fight tile to go to.

**Comment by Daniferrito** [comment]
> Its worth updating the logic to go into ALL reachable non-combat tiles immediately, before picking the potentially best fight tile to go to.

This is already what it does. It will not do a fight unless it has vision over all available non-fight tiles.

In the gif, the first fight for the first loop was forced (all three surrounding tiles were fights). All other loops only fought the boss.

It will not bother going to empty tiles that will not reveal anything. It just loses ticks without gaining anything.

**Comment by lmikulla** [comment]
Fully correct, was making assumptions based on the gifs without looking at code. 👍

**Comment by Daniferrito** [comment]
So, for conditions, what players might want is:
* Do everything (all fights, then all chests), then go to the boss
* Do all fights, then go to the boss (ignoring chests)
* Do all chests, then go to the boss (ignoring fights)
* Go directly to the boss (ignoring fights and chests)

Special: 
* Do only the chests (ignoring fights and the boss), to farm Darumaka Zen
* Do only the fights (ignoring chests and the boss), to capture a non-boss pokemon (this is probably triggered from another function, like a future "Capture all pokemons" focus)
* Do all fights, then all chests, then reset. Like the previous one, but even more niche.
* Do nothing (ignore all). This one is the only trully useless.

By the pidgeonhole principle, if we want to allow 5 posibilities, two switches are not enough. We need either a third switch or a dropdown.

I suggest to have three switches:
* Do all fights
* Collect all chests
* Kill boss

However, this means changing the previous switches meanings.
If all switches are off, the dungeon should not be startable (there is nothing to do anyway)

File: src/lib/Dungeon.js [hunk 1]
```diff
@@ -353,6 +353,56 @@ class AutomationDungeon
...
 
+            // If flash is unlocked, use it to avoid fighting every encounters
+            if (DungeonRunner.map.flash)
+            {
+                const floor = DungeonRunner.map.playerPosition.peek().floor;
+                const currentBoard = DungeonRunner.map.board.peek()[floor];
+                const unwrapedTiles = currentBoard
+                    .flatMap((row, i) => row.map((tile, j) => ({ tile, i, j })))
+                    .reduce((acc, { tile, i, j }) => ({...acc, [`${i},${j}`]: {tile, i, j}}), {});
+                const visibleUnvisitedTiles = Object.values(unwrapedTiles).filter(({ tile }) => tile._isVisible && !tile._isVisited);
```
Comment (Farigh) [inline src/lib/Dungeon.js:364]: avoid using pokeclicker internal variable, prefer using the public interface:
```suggestion
                const visibleUnvisitedTiles = Object.values(unwrapedTiles).filter(({ tile }) => tile.isVisible && !tile.isVisited);
```

File: src/lib/Dungeon.js [hunk 2]
```diff
@@ -353,6 +353,56 @@ class AutomationDungeon
...
+                const nonEnemyCells = visibleUnvisitedTiles.filter(({ tile }) => tile.type.peek() !== GameConstants.DungeonTile.enemy);
+                const bossCell = nonEnemyCells.find(({ tile }) => tile.type.peek() === GameConstants.DungeonTile.boss);
+                const chestCell = nonEnemyCells.filter(({ tile }) => tile.type.peek() === GameConstants.DungeonTile.chest);
```
Comment (Farigh) [inline src/lib/Dungeon.js:367]:
```suggestion
                const nonEnemyCells = visibleUnvisitedTiles.filter(({ tile }) => tile.type() !== GameConstants.DungeonTile.enemy);
                const bossCell = nonEnemyCells.find(({ tile }) => tile.type() === GameConstants.DungeonTile.boss);
                const chestCell = nonEnemyCells.filter(({ tile }) => tile.type() === GameConstants.DungeonTile.chest);
```

File: src/lib/Dungeon.js [hunk 3]
```diff
@@ -353,6 +353,56 @@ class AutomationDungeon
...
+
+                if (chestCell.length > 0)
+                {
+                    chestCell.forEach(({ i, j }) => this.__internal__addChestPosition({x: j, y: i, floor}));
+                }
+
+                if (this.__internal__bossPosition !== null)
+                {
+                    if ((Automation.Utils.LocalStorage.getValue(this.Settings.BossRush) === "true")
+                        || (this.__internal__chestPositions.length === DungeonRunner.maptotalChests - 1))
```
Comment (Farigh) [inline src/lib/Dungeon.js:382]: We should not consider chests if the option to skip those is enabled

File: src/lib/Dungeon.js [hunk 4]
```diff
@@ -415,6 +465,38 @@ class AutomationDungeon
...
+                unwrapedTiles[`${i - 1},${j}`],
+                unwrapedTiles[`${i - 1},${j + 1}`],
+                unwrapedTiles[`${i},${j - 1}`],
+                unwrapedTiles[`${i},${j + 1}`],
+                unwrapedTiles[`${i + 1},${j - 1}`],
+                unwrapedTiles[`${i + 1},${j}`],
+                unwrapedTiles[`${i + 1},${j + 1}`],
+            ].filter(neigh => neigh && !neigh.tile._isVisible).length;
+            return { tile, i, j, nonVisibleNeighborsCount };
+        });
```
Comment (Farigh) [inline src/lib/Dungeon.js:489]: I found this difficult to read and understand, I'm not familial with this syntaxe, is there a more readable way to write this ?

File: src/lib/Dungeon.js [hunk 5]
```diff
@@ -473,14 +555,14 @@ class AutomationDungeon
...
-        if (!this.__internal__isFirstMove)
+        if (!this.__internal__isFirstMove && !DungeonRunner.map.flash)
         {
             this.__internal__playerActionOccured = this.__internal__playerActionOccured
                                                 || !DungeonRunner.map.board()[currentFloor].every((row) => row.every((tile) => tile.isVisited));
         }
 
         // The boss was not found, reset the chest positions and move the player to the entrance, if not already there
-        if (!this.__internal__isCompleted && this.__internal__playerActionOccured)
+        if (!this.__internal__isCompleted && (this.__internal__playerActionOccured || DungeonRunner.map.flash))
```
Comment (Farigh) [inline src/lib/Dungeon.js:565]: Why are we moving back to the starting tile in that case ?

File: src/lib/Dungeon.js [hunk 6]
```diff
@@ -353,6 +353,56 @@ class AutomationDungeon
...
             }
 
+            // If flash is unlocked, use it to avoid fighting every encounters
+            if (DungeonRunner.map.flash)
+            {
+                const floor = DungeonRunner.map.playerPosition.peek().floor;
+                const currentBoard = DungeonRunner.map.board.peek()[floor];
+                const unwrapedTiles = currentBoard
+                    .flatMap((row, i) => row.map((tile, j) => ({ tile, i, j })))
+                    .reduce((acc, { tile, i, j }) => ({...acc, [`${i},${j}`]: {tile, i, j}}), {});
```
[the remark this answers is not in the document]
Comment (Farigh) [reply src/lib/Dungeon.js:363]: Maybe ?
```suggestion
                    .reduce((acc, { tile, i, j }) => 
                    { 
                        acc[`${i},${j}`] = {tile, i, j};
                        return acc;
                    }, {});
```

If I understood correctly ?

**Comment by Farigh** [comment]
Very good work, this will be a huge improvement

File: src/lib/Dungeon.js [hunk 7]
```diff
@@ -353,6 +353,56 @@ class AutomationDungeon
                 return;
             }
 
+            // If flash is unlocked, use it to avoid fighting every encounters
+            if (DungeonRunner.map.flash)
+            {
+                const floor = DungeonRunner.map.playerPosition.peek().floor;
+                const currentBoard = DungeonRunner.map.board.peek()[floor];
```
[the remark this answers is not in the document]
Comment (Daniferrito) [reply src/lib/Dungeon.js:360]: I was sure that the `peek()` is what pokeclicker was doing in its own code, but it seems it only does that in the Deals (berry, daily, gem and shards)

File: src/lib/Dungeon.js (hunk 1 above)
Comment (Daniferrito) [reply src/lib/Dungeon.js:364]: I was sure that the `isVisible` did not exist, that is why I was using the internal one. But it is indeed there

File: src/lib/Dungeon.js (hunk 5 above)
Comment (Daniferrito) [reply src/lib/Dungeon.js:565]: It was freezing if it hapened to cross a top corner without seeing the boss, but reading it again, this does not make sense. I will review if it is still needed, or if something else fixed that particular problem.

**Comment by Farigh** [comment]
Side-note: I really think it would be a good time to add some unit tests for this class

File: src/lib/Dungeon.js (hunk 4 above)
Comment (Farigh) [reply src/lib/Dungeon.js:493]: Maybe add a function `__internal_getNeighbors` and instead of building the array in-place
in the function build an array an use a loop to insert elements to it.

I'm also thinking an object might not be the best choice for `unwrapedTiles`
Instead of building it this way, couldn't you use a two-dimensional array instead ?

**Comment by Daniferrito** [comment]
Many changes.

* This should handle correctly all possible conditions combinations, both with flash and without flash.
* It should also handle correctly multi-layer dungeons with flash, even though they are not available right now (by the time flash is unlocked for a dungeon, the size has shrinked to a single layer). They will be available when the next region is available. I had to cheat to unlock flash for a normal dungeon before it shrinks to be able to test.
* This should also cover ALL requested changes, and all comments from this PR. 

I tested most posible combinations for a while, but some edge cases only happen for weird maps, so I might have missed something. I did solve all edge cases that I found.

> Side-note: I really think it would be a good time to add some unit tests for this class
I agree, but I personally dont want to reimplement all the base pokeclicker dungeon handling code in js to be able to test the automation. In my opinion it should import the real code instead of using stubs, monkeypatching the required methods as necessary (so the farming tests dont have to wait all the real hours to see if they work). However, that would be a totally different thing.
Aditionally, just copying the code brings problems with licensing, as base pokeclicker does not have a license, we cannot just copy their code to this repo (but testing could just make a pull to get the code for the tests).

**Comment by Farigh** [comment]
You're right about the tests, it seemed a good approach in the beginning, but it's becoming hard to maintain. I'll see if I can rework this to use the pokeclicker script without having to start the whole app.
I have to finish something with the desktop app first.

File: src/lib/Dungeon.js [hunk 8]
```diff
@@ -282,137 +325,279 @@ class AutomationDungeon
...
-                if (Automation.Utils.LocalStorage.getValue(this.Settings.BossRush) === "true")
-                {
-                    this.__internal__isCompleted = true;
+                    if (visibleEndTile)
+                    {
+                        this.__internal__moveToCell(this.__internal__bossPosition);
+                        // Reset current floor states before moving to the next one
+                        this.__internal__chestPositions = [];
+                        this.__internal__bossPosition = null;
+                        // Do not call DungeonRunner.nextFloor() directly, as it has no checks
```
Comment (Farigh) [inline src/lib/Dungeon.js:340]: That was the point, if we are here, we should always be in the state that would lead directly to this call (and avoid all those check overhead).

If it's not the case, we should do it manually, since resetting boss chests and boss position would make us in an undefined state

File: src/lib/Dungeon.js [hunk 9]
```diff
@@ -282,137 +325,279 @@ class AutomationDungeon
...
+        { x: 0, y: 1 },
+        { x: -1, y: 0 },
+        { x: 1, y: 0 },
+    ];
+
+    /**
+     * @brief Returns the list of neighbouring cells
+     *
+     * This is similar to the map.nearbyTiles() function, but it returns the coordinates as well
+     * It cant be as short as the map.nearbyTiles() function we dont have the ?. operator in js to check for the bounds
```
Comment (Farigh) [inline src/lib/Dungeon.js:565]: Are you sure about that ? 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

I tried in the firefox console and it worked as intended.

File: src/lib/Dungeon.js [hunk 10]
```diff
@@ -441,19 +634,7 @@ class AutomationDungeon
...
-                    }
-
-                    if (tile.type() === GameConstants.DungeonTile.boss)
-                    {
-                        // Only tag the dungeon as completed if it's not the first move or any tile is visited apart from the entrance
-                        this.__internal__isCompleted = (!this.__internal__isFirstMove) || DungeonRunner.map.board()[currentFloor].some(
-                            (row) => row.some((tile) => tile.isVisited && (tile.type() !== GameConstants.DungeonTile.entrance)));
-                        this.__internal__bossPosition = currentLocation;
-                    }
+                    this.__internal__markCell({ ...currentLocation, tile });
```
Comment (Farigh) [inline src/lib/Dungeon.js:637]: any reason to do it only if the player has access to it in this case ? 
The other calls are not protected this way

**Review by Farigh — COMMENTED** [review]
I made some comments and pushed some quick fixes

File: src/lib/Dungeon.js [hunk 11]
```diff
@@ -264,15 +296,38 @@ class AutomationDungeon
             if (this.__internal__isFirstMove)
             {
                 this.__internal__ensureTheBoardIsInAnAcceptableState();
+                if (DungeonRunner.map.flash)
+                {
+                    this.__internal__markNearbyTiles();
+                }
```
Comment (Farigh) [inline src/lib/Dungeon.js:302]: wouldn't this be already set bu the `this.__internal__ensureTheBoardIsInAnAcceptableState();` call ?

**Review by Daniferrito — COMMENTED** [review]
Testing review to see if It reopens the suggestions

File: src/lib/Dungeon.js (hunk 8 above)
Comment (Farigh) [reply src/lib/Dungeon.js:342]: That's fine as-is for now, if any perf overhead is found their, will think about it at that moment

File: src/lib/Dungeon.js (hunk 10 above)
Comment (Farigh) [reply src/lib/Dungeon.js:615]: Looks good to me then

**Review by Farigh — APPROVED** [review]
I rebased and reworked the history to have more meaningful commits.
I renamed the SkipBoss internal setting to reflect the new behaviour

Thanks for your great work

File: src/lib/Dungeon.js [hunk 12]
```diff
@@ -6,14 +6,15 @@ class AutomationDungeon
     static Settings = {
                           FeatureEnabled: "Dungeon-FightEnabled",
                           StopOnPokedex: "Dungeon-FightStopOnPokedex",
-                          BossRush: "Dungeon-BossRush",
-                          SkipChests: "Dungeon-SkipChests"
+                          AvoidEncounters: "Dungeon-AvoidEncounters",
```
Comment (Daniferrito) [inline src/lib/Dungeon.js:9]: Should not have changed the setting name, as it makes the setting reset to default. Not a big issue, but try not to change this as it messes with people's default.

If this needs to be done anyways, set the default from the old value from the old name.

Comment (Farigh) [reply src/lib/Dungeon.js:9]: I did it on purpose, since the meaning changed a bit from the previous behaviour
